### ISSUE TITLE
Notebook controller is only checking the status of one (the first) container

### ISSUE THREAD

**Issue by gilbeckers** [kubeflow/kubeflow#5282]
/kind bug

**What steps did you take and what happened:**
When you add a new notebook server, the status of the newly added notebook becomes ready to soon. Namely, if you click on connect, a "no healthy upstream" message shows the first seconds. I tried to fix this by adding a readiness probe to the notebook container (via a custom admission controller). However, the problem stayed the same: the green check mark appears under status too soon.

**What did you expect to happen:**
Correct behavior of the status icon of the notebook server in the Kubeflow dashboard.

**Anything else you would like to add:**
I think I know what the problem is and why the status is wrongly reported in the Kubeflow dashboard: The notebook controller is only checking the status of the first container https://github.com/kubeflow/kubeflow/blob/c4717d84b6d927b7cf2305511002094e9d19f5a4/components/notebook-controller/controllers/notebook_controller.go#L210 

As a result, it's only taking the status of the istio-proxy container in consideration, and the status of the notebook container is ignored. I think this is a little bug, as the controller should definitely check the status of all containers, and not only the first one.

I'm new to this community, but if it's ok I would like to fix this and add a new pull request.

**Environment:**

- Kubeflow version: (version number can be found at the bottom left corner of the Kubeflow dashboard): 1.0.2
- kfctl version: (use `kfctl version`): 1.0.2
- Kubernetes platform: (e.g. `minikube`): minikube
- Kubernetes version: (use `kubectl version`): client v1.19.0 , server v1.18.3
- OS (e.g. from `/etc/os-release`): Ubuntu 18.04

**Comment by gilbeckers** [comment]
**Update:**

I managed to fix the wrongly reported status of a new created notebook server in the kubeflow dashboard. 
Now, the status of nb server becomes ready only after the two containers (isto-proxy and notebook) are both Running and Ready. 

In order to fix this, two changes were needed: 
- The netbook controller was only checking the Status of the first container, resulting in only the istio container being taken into account (as stated in the first comment above). I changed this so that now all containers in the pod are considered.
https://github.com/gilbeckers/kubeflow/blob/db06cc850e9ef8fa602ba825444099cca2ee0f95/components/notebook-controller/controllers/notebook_controller.go#L217

- The second change is something I discovered only later while examining the code more in-depth: At this moment, the notebook controller looks only at the Status field of the container. However, there is also a Ready flag which can still be false while the Status of the container is already Running. I changed this so that now the pod status only becomes Running if all containers in the pod are Running and have their Ready flag set to true.

In addition, I added a readiness probe to the nb container. Until now only the isto container has a readinessProbe. We decided to add this probe using a custom admission controller because we didn't want to mess with the PodDefaults. I'm not 100% sure if the extra added readiness probe is strictly necessary, as the status is also correct reported in the kf dashoboard in case no probe is added. (Also not sure if this chunk should be part of a possible pull request)

**Comment by gilbeckers** [comment]
@MartinForReal  I'm not sure if the linked pull request will completely fix this issue, as the bug of only looking at the status field of one container still remains.

**Comment by MartinForReal** [comment]
@gilbeckers the state of the container will be ready only after the container has passed all of the health checks. 
As the code here 

https://github.com/gilbeckers/kubeflow/blob/db06cc850e9ef8fa602ba825444099cca2ee0f95/components/notebook-controller/controllers/notebook_controller.go#L248

The controller will check if all of the containers are up and running. So by adding a liveness probe, we can fix this problem. 

How do you think?

**Comment by gilbeckers** [comment]
@MartinForReal I will clarify my current situation. In my notebook pod I have 2 containers running: a istio proxy container and a container that runs the notebook image.

The problem with the current notebook controller in this situation is twofold:
- The nb controller only checks the state of the first container, this is hard coded: 
https://github.com/kubeflow/kubeflow/blob/7f5e242f4671f453eaee72e4319a5f7f190f7725/components/notebook-controller/controllers/notebook_controller.go#L210
As a result the status of the second container (the one running the notebook image) is not being taken into account.
- Besides, in the current implementation of the nb controller only the State field of the container (`pod.Status.ContainerStatuses[i].State`) is checked. However there is also a Ready flag (`pod.Status.ContainerStatuses[i].Ready`) which should be checked on as well. This is because the container can be in the Running state while it's Ready field is still set to false.

You say that the container will be ready only after the container has passed all health checks, and by adding a liviness/readiness probe we can fix this problem. This is true, however the problem remains that the current notebook controller doesn't care about the ready fields of the containers and only looks at the status field of the first container.

**Comment by kimwnasptd** [comment]
@gilbeckers indeed the Controller doesn't set the State of the CR correctly. In many cases the first container is the istio-sidecar, which results in an incorrect state being shown.

So the question becomes which one of the containers' `State` should the Controller choose to depict in the Status of the CR?
I propose that we select the container whose name matches the name of the notebook, since the Jupyter web app will put the notebook's name as a value to the [name of container the user configured](https://github.com/kubeflow/kubeflow/blob/master/components/jupyter-web-app/backend/kubeflow_jupyter/common/yaml/notebook.yaml#L14). 

If no container matched this condition then lets have the controller to throw a big warning to notify the users that it couldn't identify the "notebook" container.

@gilbeckers WDYT?

**Comment by kimwnasptd** [comment]
Also, I think the Jupyter web app should be using the `readyReplicas` field in the status to decide whether the Notebook is running or not and not check the Container State

**Comment by thesuperzapper** [comment]
@kimwnasptd @gilbeckers shouldn't the state of the Notebook be an `and` type operation of ALL the container states?

Example:
`Jupyter = Healthy`, `istio = Healthy` --> `Healthy`
`Jupyter = Healthy`, `istio = Unhealthy` --> `Unhealthy`
`Jupyter = Unhealthy`, `istio = Healthy` --> `Unhealthy`
`Jupyter = Unhealthy`, `istio = Unhealthy` --> `Unhealthy`

**Comment by gilbeckers** [comment]
@thesuperzapper at this moment I implemented it as an `and` operation indeed, so I was thinking the same as you. In that way, you just can loop over all containers, without caring about the actual names of containers.

About the suggestion of @kimwnasptd : I'm no sure, but maybe it's not the best idea to make the logic jupyter specific (as we want to abstract the way apps are deployed in Kubeflow, as already stated in #5292 )

**Comment by kimwnasptd** [comment]
> @kimwnasptd @gilbeckers shouldn't the state of the Notebook be an and type operation of ALL the container states?

I don't think this is a good idea. 
The CR's state that we mention is a [`ContainerState`](https://github.com/kubeflow/kubeflow/blob/7f5e242f4671f453eaee72e4319a5f7f190f7725/components/notebook-controller/api/v1/notebook_types.go#L44). As also noted in the [docs](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.19/#containerstatus-v1-core) the `ContainerState` should depict the state of **one** container. So I think we shouldn't diverge from this pattern and treat the `ContainerState` as the logical AND of all the Pod's containers.

> Example:
Jupyter = Healthy, istio = Healthy --> Healthy
Jupyter = Healthy, istio = Unhealthy --> Unhealthy
Jupyter = Unhealthy, istio = Healthy --> Unhealthy
Jupyter = Unhealthy, istio = Unhealthy --> Unhealthy

@thesuperzapper I'm not sure I follow here. Doesn't a `ContainerState` only take `running`, `terminated` or `waiting` values?
[ I'm mentioning `ContainerState` because in your comment above you mentioned container state. Did you mean something else? ]

The pattern that you mentioned is depicted in the `readyReplicas` in our CR, which mirrors the underlying StatefulSet's `readyReplicas`. So if all of the containers are `Ready` then the entire Pod will become ready [ there's a really good doc that explains the relationship between Pod <> Container readiness in [this k8s enhancement](https://github.com/kubernetes/enhancements/blob/master/keps/sig-network/0007-pod-ready%2B%2B.md) ]

So to sum up I think we should:
1. Stick to showing only the Notebook's container `ContainerState` to the CR's `ContainerState`
2. Use the `readyReplicas` to determine if the Pod is ready

> About the suggestion of @kimwnasptd : I'm no sure, but maybe it's not the best idea to make the logic jupyter specific (as we want to abstract the way apps are deployed in Kubeflow, as already stated in #5292 )

@gilbeckers why do you think my proposal is jupyter specific? The selecion of the Notebook container is based on the name of the contaner. How do you think it would limit us for supporting different workloads?

**Comment by thesuperzapper** [comment]
bump
